server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java

```diff
@@ -554,6 +580,11 @@ public SystemIndexDescriptor getDescriptorCompatibleWith(Version version) {
         return null;
     }
 
+    public static int getDescriptorHash(String mappings, Settings settings) {
+        Map<String, Object> convertedMappings = XContentHelper.convertToMap(XContentType.JSON.xContent(), mappings, false);
+        return Objects.hash(convertedMappings, settings);
```

Comment:
Two comments on the hash:
1. I'm happy to use something other than `Objects::hash` here, but it's very convenient for a first cut, since we can get a hash code for the mapping and setting objects.
2. Do we need to store hardcoded hashes? It's a lot of file changes and additional friction for modifying any system index mappings or settings. The advantage is that we can nudge users into adopting the "prior index descriptor" system that's been underused so far, and provide clearer guidance on when and how to modify a system index descriptor.

Reply:
I believe the hash values of all json core types are guaranteed by the JVM (@ChrisHegarty are they?), but if there's a more 'library' method somewhere rather than relying on a JVM intrinsic that might read better

Reply:
As per one of my other comments I think it would be a good idea to remove the version meta key from the map before creating the hash. That way the hardcoded constants won't become invalid when the main branch version is bumped.

Reply:
The `Map::hashCode` contract is agnostic of entry order.

Reply:
> I believe the hash values of all json core types are guaranteed by the JVM

This is not the case. The JSON types here are from Jackson, right? Not the JDK.  If so, then their hashCode contract is specified and controlled by the Jackson maintainers (rather than the JDK or JVM).

Reply:
Well, the doc for [`Map.hashcode`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Map.html#hashCode()) specifies the algorithm to use, which is unordered. So if the jackson maps are doing something different that's breaking spec